### DISCUSSION TITLE
Dioxus 0.6 Roadmap

### DISCUSSION THREAD

**Discussion by jkelleyrtp** [DioxusLabs/dioxus#2223] [Ideas]
# Dioxus 0.6 Roadmap

This release focus on polishing existing APIs, improving developer experience, upgrading fullstack, and the initial release of a component library.

### Fullstack basics

*Goal: expose as much as axum as possible to make fullstack more complete. Cater to real-world usecases like file uploads, SSG, streaming, and the dioxus-helmet replacement.*

- Head element 
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1169
    - Merging of `head {}` element in renderers
- Cleaning up assets
- File uploads - Streaming bytes works, integrate with the dioxus html crate 
    - [ ]  https://github.com/DioxusLabs/dioxus/issues/1333
- File Interface Rework to Support Streaming 
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1920
- Streaming HTML - #2365
- Websockets - Work with leptos server fn, unknown amount of work
- SSG - Almost done, we use this on the docsite already 
    - [x] https://github.com/DioxusLabs/dioxus/issues/2006
- Progressive forms - Have a prototype
- Restore finishing suspense - #2365
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1335
- Bundle splitting - If we have time 
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1578
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1191
- Examples for auth/token flow, fix bug in router related to tokens
    - [x] https://github.com/DioxusLabs/dioxus/issues/1994
    - [x] https://github.com/DioxusLabs/dioxus/issues/1847
- Tier 3 LiveView (downgrade our support)

### Component Library

*Goal: a first-class component library for dioxus that takes into account mobile. Starting with solid foundations (like typesafe interfaces) and then exposing useful components on top. Documentation added to docsite.*

https://github.com/DioxusLabs/dioxus/issues/1483

- mobile-oriented components
- a base like radix-ui to build on top of
- potentially derive or context based APIs - requires some experimentation
- examples on the docsite
- maybe alpha.0
- stress-testing of the devtools
- rename std to sdk
- dogfood with real app
- Inspiration:
    - https://mui.com/ (good block abstractions)
    - https://tamagui.dev/ (has interest in mobile)
    - https://ionicframework.com/docs/components (components for cross-platform)
    - [https://ui.shadcn.com/](https://ui.shadcn.com/) - these just look good
- Component ideas
    - Take a look at https://ui.shadcn.com/docs/components/context-menu - lots of ideas here
- A playground for iterating and exporting components?
    - We could run a little VPS like yew with browser-based hotreloading

### Polishes - focus on CLI and documentation

*Fix the various rough edges that have cropped up during 0.5 testing. Make hotreload more reliable across the board. Plugins, crash screens, smoother overall experience, better docs.*

- More hotreloading
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1964
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1867
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1955
    - [ ] https://github.com/DioxusLabs/dioxus/issues/2032
    - [ ] https://github.com/DioxusLabs/dioxus/issues/2079
    - [ ] https://github.com/DioxusLabs/dioxus/issues/2077
- hotreloading of prop literals (IE attributes like class)
- hotreloading of formatted strings
- serve old version of site while old is building
- crash/fail to compile screens
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1989
- plugins
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1192
- inline mobile tooling (?) - time dependent
    - [ ] https://github.com/DioxusLabs/dioxus/issues/2091
    - [ ] https://github.com/DioxusLabs/dioxus/issues/2097
- videos & tutorials
    - https://javalin.io/tutorials/
    - laravel’s tutorials
- (experimental) hotreloading of rust? cranelift?

### Native

*Making stylo-dioxus interactive and at feature-parity with blitz. Replacing blitz with stylo-dioxus. Improvements to taffy that are necessary for html layout (like inline and style abstraction)*

- [ ] native dom representation
- [ ] get regular dioxus plumbing into native (hotreload, events, devtools, launch)
- [ ] calc support in taffy
- [ ] “style” abstraction in lieu of `Style` object in taffy
- [ ] split apart dioxus-desktop to be used for stylo-dioxus
- [ ] mixed content progress

### Bug Fixes/tweaks

*A variety of bug fixes and tweaks that we want to do during this release that would normally get caught in a backlog.*

- prevent_default is properly sync now desktop/mobile (conditional prevent default)
    - [ ] https://github.com/DioxusLabs/dioxus/issues/1662
- [ ] virtual file API in lieu of FileEngine
- [ ] files/assets/paths not working across platforms properly (loading images, etc)
- [ ] enable all doctesting #2460
- docs and recipes, more photos, more guides
  - [ ] better inline docs/examples #2460
- More stuff to come - whatever we can pull of the backlog that fits

### CI/CD

*Improvements to our continuous release process*

- Bundle sizes, perf on a floating PR
- Floating PR that tracks closed PRs since last release, semver compatibility, etc
- Autorelease workflow for patch versions that evan can merge
- Maybe improve how we release the CLI since we just stuff all the CLIs in the single release
- Bubble up the “status” onto a page on the website - maybe reading the open floating PR

**Comment by gbj** [comment]
> - File uploads - Streaming bytes works, integrate with the dioxus html crate
> - Parallel streaming server functions - Stream ID to the client and resume, unknown amount of work
> - Websockets - Work with leptos server fn, unknown amount of work

Let me know if there's any upstream `server_fn` support that can be helpful here, and I'd love to have any of that as PRs. The `server_fns_axum` example in our repo has file upload examples. I'm not thrilled with duplexing/progress requiring extra machinery so happy to work together here. 

> Progressive forms - Have a prototype

Likewise [the form implementation we have now](https://github.com/leptos-rs/leptos/blob/leptos_0.7/leptos/src/action_form.rs) is very straightforward given the server fn POST approach and just using `FormData` to send a fetch. Feel free to borrow and share any improvements!

> Bundle splitting - If we have time

In exchange I'm looking forward to stealing whatever you do here 😀 My sense is that Evan knows a lot more about WASM than I do. 

Needless to say all the rest looks amazing. You're crushing it!

**Comment by ochrons** [comment]
Would love to see the roadmap features fleshed out in more detail, including problem description and a "definition of done", to make them more concrete. As examples, some features I would like to see for fullstack that are currently missing or difficult to use.

## Using a database in server functions

#### Problem
Calling database from a server function requires a way to pass the database pool to the function. In Axum this can be done via the State extractor feature, storing the DB pool in the State and extracting it in the function. This is currently not possible for Dioxus server functions (or at least not documented how to do it). A workaround is to store the state in a `static OnceLock` global variable.

#### Why is this important?
For any realistic web application, fetching data in server functions is a must. The current approach is non-idiomatic, not documented and cumbersome.

#### Definition of done
As a developer I can extract the Axum State object in the server function to make DB calls.

This feature is documented on the docsite and is used in an example application.

## Compile and deploy the server on different architecture

#### Problem
I use x64 based Windows computer but would need to run the application server on AArch64 Linux cloud server. I would like to build a Docker image for this platform on my computer, or in CI. If it is possible to build the fullstack server with a different target architecture, this is not documented.

#### Why is this important?
Being able to run the server on different architecture from the one that the developer or CI is using, is often needed as cloud providers support both ARM and x64 architectures.

#### Definition of done
As a developer I can define a build script using either `cargo` or `dx` that builds the server binary to the target architecture I need. 

This feature is documented on the docsite, including how to build a Docker container with the fullstack app.

## SSR with page specific metadata in `head`

#### Problem
My website uses SSR to improve SEO, but currently there is no way to populate important metadata in the `head` section.

#### Why is this important?
SSR loses a lot of its potential if the page metadata cannot be set to improve SEO, or to enable embeddings.

#### Definition of done
As a developer I can write Dioxus SSR code that fetches page specific data from a database and populates metadata elements in the `head` section using this data during the server side rendering step. The metadata is also updated accordingly on client side when navigating to different pages.

This feature is documented on the docsite and an example is provided.

**Reply by benlad** [reply]
+1 Using a database in server functions

**Reply by dxps** [reply]
I made an example for this particular case since I was also very interested if and how it is possible.
Definitely it wasn't a smooth road to get there. Please [find it here](https://github.com/dxps/dioxus_playground/tree/main/dioxus-05-fullstack-routing-axum-pgdb). Hope it helps!

**Reply by ealmloff** [reply]
https://github.com/DioxusLabs/dioxus/pull/2635 adds head elements that integrate with fullstack and https://github.com/DioxusLabs/dioxus/pull/2515 makes it easier to provide context to server functions in fullstack. The docsite will be updated with these features with the release.

It is much easier to build macos/ios on a mac device so cross platform builds will probably be a github action instead of a cross platform build. A build action is planned for 0.7 (pulled out + improved from [here](https://github.com/floneum/floneum/blob/main/.github/workflows/nightly.yml)).

**Comment by wheregmis** [comment]
First of all, Rust is Amazing and Dioxus is making me love Rust every single day.

Talking about Fullstack Application on Dioxus. As the project grows its super hard to go just use
`dx server --platform fullstack`

As it compiles web and server both and since hot reloading on dioxus is really amazing but just a small project with auth, few database libraries will make it frustrating to work. And majority of the time i feel myself working on styling frontend and sometimes alot of controller or abstraction layers to make use of it on server function. Because on WASM deployment of dioxus fullstack is really enclosed by dioxus i tried using 

https://github.com/DioxusLabs/dioxus/tree/main/packages/fullstack/examples/axum-desktop/

This example as a base where i have seperate client and server (exposing axum) and making it work on web. So if i am working on frontend i can just do

`dioxus_web::launch::launch_cfg(app, dioxus_web::Config::default());`

`dx serve --platform web --features web`

Thus just compiling the frontend and having hot reload on (Really helps me to excel development on Frontend Side)

And if i want to run things on server I can do

`cargo run --bin server --features server`

We can just do

 ```rust
 use server_fn::axum::register_explicit;
        register_explicit::<PostServerData>();
        register_explicit::<GetServerData>();

        let app = Router::new()
            // Server side render the application, serve static assets, and register server functions
            .register_server_fns();

        // run it
        let addr = std::net::SocketAddr::from(([127, 0, 0, 1], 8000));
        let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();

        axum::serve(listener, app.into_make_service())
            .await
            .unwrap(); 
 ```
 
 This way its seperated from frontend (This doesnt make it fullstack but if we fine tune it using some features we can make it work to serve dioxus application as well)
 
 So in minimal example it will be like
 Client.rs
 ```rust
 
    #[cfg(feature = "web")]
    // Hydrate the application on the client
    dioxus_web::launch::launch_cfg(app, dioxus_web::Config::new().hydrate(true));

    #[cfg(feature = "just_web")]
    dioxus_web::launch::launch_cfg(app, dioxus_web::Config::default());
}
 ```
 
 Server.rs
  ```rust
  
#[tokio::main]
async fn main() {
    // build our application with some routes

    use axum::http::Method;
    use tower_http::cors::CorsLayer;

    #[cfg(not(feature = "server"))]
    {
        // TODO: We will use this to build desktop application

        use server_fn::axum::register_explicit;
        register_explicit::<PostServerData>();
        register_explicit::<GetServerData>();

        let app = Router::new()
            // Server side render the application, serve static assets, and register server functions
            .register_server_fns();

        // run it
        let addr = std::net::SocketAddr::from(([127, 0, 0, 1], 8000));
        let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();

        axum::serve(listener, app.into_make_service())
            .await
            .unwrap();
    }
    #[cfg(feature = "server")]
    {
        let origins = [
            "http://localhost:8000".parse().unwrap(),
            "http://127.0.0.1:8000".parse().unwrap(),
        ];
        let cors = CorsLayer::new()
            // allow `GET` and `POST` when accessing the resource
            .allow_methods([Method::GET, Method::POST])
            // allow requests from any origin
            .allow_origin(origins);
        let app = Router::new()
            // Server side render the application, serve static assets, and register server functions
            .serve_dioxus_application(ServeConfig::builder().build(), || VirtualDom::new(app))
            .await
            .layer(cors);

        // run it
        let addr = std::net::SocketAddr::from(([127, 0, 0, 1], 8000));
        let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();

        axum::serve(listener, app.into_make_service())
            .await
            .unwrap();
    }
}
```

lib.rs
```rust
use dioxus::prelude::*;

pub fn app() -> Element {
    let mut count = use_signal(|| 0);
    let mut text = use_signal(|| "...".to_string());

    rsx! {
        h1 { "High-Five counter: {count}" }
        button { onclick: move |_| count += 1, "Up high!" }
     
    }
}
```
  
 This way we can enjoy full dioxus fullstack with dx serve --platform fullstack
 Just work on Web UI (Complilation wont be any issue here) with dx serve --platform web --features just_web
 Just work on server using cargo run --bin server --features core
 
 Sorry for confusing comment but maybe we can enhance this experience natively on dioxus with some feature flags or seperate commands lol 😆

**Reply by gpoblon** [reply]
While it would indeed be great for testing (with mockup requests when building frontend)
This also makes a lot of sense in a production environment.
+1 for a similar solution allowing to build the same codebase either as fullstack or with separated front/back.
But you might have already been working on it, where are we at now @ealmloff?

**Reply by ealmloff** [reply]
> But you might have already been working on it, where are we at now @ealmloff?
Fullstack applications require some synchronization between the frontend and backend. They both need to render the same HTML to make hydration work and they both need to agree on what endpoints exist and the APIs of those endpoints.

It sounds like the main benefit of separated builds is compile times. Once in place binary patching is working, most of the compile time benefits should be possible in the existing fullstack system. If a change requires new data from the server, you need to rebuild both the server and client. If it only requires changes to the frontend, then the frontend wasm can be patched in place without rebuilding the server

**Reply by wheregmis** [reply]
> > But you might have already been working on it, where are we at now @ealmloff?
> > Fullstack applications require some synchronization between the frontend and backend. They both need to render the same HTML to make hydration work and they both need to agree on what endpoints exist and the APIs of those endpoints.
> 
> It sounds like the main benefit of separated builds is compile times. Once in place binary patching is working, most of the compile time benefits should be possible in the existing fullstack system. If a change requires new data from the server, you need to rebuild both the server and client. If it only requires changes to the frontend, then the frontend wasm can be patched in place without rebuilding the server

+1 
And for server, we can use cranelift to speed up the compilation until we have Native Rust Binary Patching. (For my purpose and dependencies its working well)
